File: Exámenes/2020/Febrero/Haskell/DemoFordFulkerson.hs
-------------------------------------------------------------------------------
-- Ford-Fulkerson Algorithm. Maximal flow for a weighted directed graph.
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module FordFulkerson where

import DataStructures.Graph.FordFulkerson
import DataStructures.Graph.WeightedDiGraph
import DataStructures.Graph.WeightedDiGraphBFT

{- Some examples and expected results:

> maxFlowPath [E 'A' 3 'B', E 'C' 1 'D', E 'A' 7 'D']
1

> updateEdge 'a' 'b' 7 [E 'a' 5 'c', E 'a' 4 'b', E 'b' 3 'a']
['a'-5->'c','a'-11->'b','b'-3->'a']

> updateEdge 'a' 'b' (-4) [E 'a' 5 'c', E 'a' 4 'b', E 'b' 3 'a']
['a'-5->'c','b'-3->'a']

> updateEdge 'a' 'z' 23 [E 'a' 5 'c', E 'a' 4 'b', E 'b' 3 'a']
['a'-5->'c','a'-4->'b','b'-3->'a','a'-23->'z']

> updateEdges [E 'a' 4 'z', E 'a' 3 'b', E 'c' 1 'd'] 2 [E 'a' 1 'b', E 'b' 2 'c', E 'c' 3 'd']
['a'-3->'b','b'-2->'c','c'-5->'d','a'-2->'z']

-}

g :: WeightedDiGraph String Integer
g = mkWeightedDiGraphSuc ["S","A","B","C","D","E","F","T"] sucs
    where
        sucs "S" = [("C",4) ,("B",8), ("A",4)]
        sucs "A" = [("D",3), ("F",7)]
        sucs "B" = [("E",9)]
        sucs "C" = [("E",5), ("F",2)]
        sucs "D" = [("T",6)]
        sucs "E" = [("T",9)]
        sucs "F" = [("T",5)]
        sucs "T" = []
        
{- Some examples and expected results:

> let list = fordFulkerson g "S" "T"
> list
["F"-5->"T","A"-3->"F","S"-4->"A","E"-9->"T","B"-8->"E","S"-8->"B",
"C"-2->"F","S"-3->"C","C"-1->"E","D"-1->"T","A"-1->"D"]

> maxFlow list "S"
15

> maxFlowMinCut g "S" "T" ["S","A","F"]
15

-}


g2 :: WeightedDiGraph String Integer
g2 = mkWeightedDiGraphSuc ["S","A","B","C","D","T"] sucs
    where
        sucs "S" = [("A",10), ("C",10)]
        sucs "A" = [("B",4), ("C",2), ("D",8)]
        sucs "B" = [("T",10)]
        sucs "C" = [("D",9)]
        sucs "D" = [("B",6), ("T",10)]
        sucs "T" = []
        
{- Some examples and expected results:
>   
> list2
["D"-10->"T","C"-9->"D","S"-9->"C","A"-6->"D","S"-10->"A",
"B"-9->"T","A"-4->"B","D"-5->"B"]

> maxFlow list2 "S"
19

> maxFlowMinCut g2 "S" "T" ["S","A","B","C","D"]
19

-}

File: Exámenes/2020/Febrero/Java/src/DemoFordFulkerson.java
import dataStructures.graph.*;
import dataStructures.list.LinkedList;
import dataStructures.list.List;
import dataStructures.set.HashSet;
import dataStructures.set.Set;

public class DemoFordFulkerson {

    public static void main(String[] args) {

        WeightedDiGraph<String, Integer> g = new WeightedDictionaryDiGraph<>(); 

        g.addVertex("S");
        g.addVertex("A");
        g.addVertex("B");
        g.addVertex("C");
        g.addVertex("D");
        g.addVertex("E");
        g.addVertex("F");
        g.addVertex("T");

        g.addDiEdge("S", 4, "A");
        g.addDiEdge("S", 8, "B");
        g.addDiEdge("S", 4, "C");
        g.addDiEdge("A", 3, "D");
        g.addDiEdge("A", 7, "F");
        g.addDiEdge("B", 9, "E");
        g.addDiEdge("C", 5, "E");
        g.addDiEdge("C", 2, "F");
        g.addDiEdge("D", 6, "T");
        g.addDiEdge("E", 9, "T");
        g.addDiEdge("F", 5, "T");

        List<WDiEdge<String, Integer>> path = new LinkedList<>();
        path.append(new WDiEdge<>("A",3,"B"));
        path.append(new WDiEdge<>("B",1,"D"));
        path.append(new WDiEdge<>("D",3,"C"));
        System.out.println("Maximal flow in path is "+FordFulkerson.maxFlowPath(path));

        List<WDiEdge<String, Integer>> edges = new LinkedList<>();
        edges.append(new WDiEdge<>("A",2,"B"));
        edges.append(new WDiEdge<>("A",5,"C"));
        edges.append(new WDiEdge<>("B",6,"D"));
        edges.append(new WDiEdge<>("D",1,"E"));
        edges.append(new WDiEdge<>("B",4,"A"));
        System.out.println(FordFulkerson.updateEdge("A","C", 3,edges));
        System.out.println(FordFulkerson.updateEdge("D","B", 2,edges));

        System.out.println(g);

        List<WDiEdge<String, Integer>> wedges = g.wDiEdges();

        FordFulkerson<String> ff = new FordFulkerson<>(g, "S", "T");
        System.out.println("Edges in maximal flow graph: " + ff.getSolution());
        System.out.println("Maximal flow rate = " + ff.maxFlow());
		
        Set<String> set = new HashSet<>();
        set.insert("S");
        set.insert("A");
        set.insert("F");
        System.out.println("Maximal flow rate using set " + set + " is " + ff.maxFlowMinCut(set));
    }
}

/* Expected output:

Maximal flow in path is 1
LinkedList(A-2->B,A-8->C,B-6->D,D-1->E,B-4->A)
LinkedList(A-2->B,A-5->C,B-6->D,D-1->E,B-4->A,D-2->B)
WeightedDictionaryDiGraph(vertices=(A,B,C,D,E,F,S,T), edges=(A-3->D,A-7->F,B-9->E,C-2->F,C-5->E,D-6->T,E-9->T,F-5->T,S-4->A,S-4->C,S-8->B))
Edges in maximal flow graph: LinkedList(S-4->A,A-3->D,D-3->T,A-1->F,F-3->T,S-4->C,C-2->F,C-2->E,E-9->T,S-7->B,B-7->E)
Maximal flow rate = 15
Maximal flow rate using set HashSet(A,F,S) is 15
*/
